pkg/endpoint/endpoint.go

```diff
@@ -2403,3 +2433,61 @@ func (e *Endpoint) IPs() []net.IP {
 func (e *Endpoint) InsertEvent() {
 	e.getLogger().Info("New endpoint")
 }
+
+// must be called with endpoint mutex locked.
```

Comment:
:+1: So this is what makes it safe to have this run synchronously from an API handler thread as well as from a controller.